src/main/java/net/minecraftforge/common/command/SelectorHandler.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.common.command;
+
+import java.util.List;
+
+import net.minecraft.command.CommandException;
+import net.minecraft.command.ICommandSender;
+import net.minecraft.entity.Entity;
+
+/**
+ * Handler for custom types of selectors registered with {@link SelectorHandlerManager}
+ */
+public abstract class SelectorHandler
+{
+    /**
+     * Returns a {@link List} of {@link Entity Entities} of class {@code targetClass} ({@code T}) represented by {@code token}
+     * 
+     * @param sender The {@link ICommandSender} that initiated the query
+     */
+    public abstract <T extends Entity> List<T> matchEntities(ICommandSender sender, String token, Class<? extends T> targetClass) throws CommandException;
+
+    /**
+     * Returns whether the selector string potentially matches multiple entities
+     */
+    public abstract boolean matchesMultiplePlayers(String selectorStr) throws CommandException;
+
+    /**
+     * Returns whether the string represents a selector <br>
+     * <b>Note:</b> Returning {@code false} does not prevent {@link #matchEntities this.matchEntities} from being called. It is recommended to not override this method and
+     * simply throw an exception there <br>
+     * <b>Note:</b> If {@code selectorStr} could be a valid entity name, this method should return {@code false} and {@link #matchEntities this.matchEntities} should return an empty list instead of throwing in case {@code selectorStr} is not a valid selector <br>
+     * <b>Note:</b> Mostly for legacy reasons.
+     * (Since {@link net.minecraft.command.EntitySelector#matchEntities(ICommandSender, String, Class) EntitySelector.matchEntities}
+     * returns an empty list when the string is illformed instead of throwing an exception (as is indirectly done when this method returns {@code false})
+     */
+    public boolean isSelector(String selectorStr)
```

Comment:
To be fair, the usage of `isSelector` is itself very confusing ;)
This is what we know about the vanilla usage of the method (maybe you have an idea on how to write that down, as we've already tried for more than an hour):
There are a few places where the following logic is more or less implemented (this example is from `CommandBase.getEntityList`):
* If `isSelector` returns `true`, `matchEntities` is called (which throws if a syntax error is encountered)
* If not, the string is passed to `getEntity`, which calls `matchOneEntity`
  * There, `matchEntities` is called again, which obviously fails (by returning an empty list, and *not* by throwing)
  * Then, the string is interpreted as player name, UUID, ...
  * If everything fails, an exception is thrown

We need to override it in order to ensure that `matchEntities` is called for custom prefixes.

As you can see, there are two possible reactions when the syntax is invalid:
* Throw an exception. This is easy to implement and works most of the time (that's why the doc suggests it). This has two possible downsides:
  * It breaks when the string could be a player name/UUID (for reasonable prefixes, this shouldn't be the case)
  * In case of the `/say` command, an error is shown instead of the unparsed selector (this happens in vanilla for any syntax error that isn't caught by the regex)
* Return `false` in `isSelector` *and* an empty list in `matchEntitites`. This however causes the whole mechanism (matching the prefix and checking the syntax) to be called twice (unnecessarily) and requires more care from the modder.

We thought it's easiest to default to `true` and change the implementation for the vanilla case. This is what the doc tries to explain briefly (as the method would look very strange otherwise).

We see three options:
* Improve the doc (we would be very happy about any suggestions ;) )
* Require vanilla behaviour from the modder (i.e. check whether the string seems reasonable (whatever that means for that specific selector) and return an empty list)
* Return `true` whenever a prefix is encountered and patch vanilla to work with that (this behaves differently for the `/say` command in a few cases). This would allow us to remove the method entirely from `SelectorHandler`

What would you suggest?